curta/src/machine/hash/sha/algorithm.rs

```diff
@@ -0,0 +1,511 @@
+                Some(reg_cycle_length),
+            );
+        }
+        for i in num_real_rounds..num_rounds - 1 {
+            builder.store(
+                &is_dummy_slice.get(i),
+                one,
+                &Time::zero(),
+                Some(reg_cycle_length),
+            );
+        }
+        let last_round_reg = builder.constant(&B::Field::from_canonical_usize(length_last_round));
+        builder.store(
+            &is_dummy_slice.get(num_rounds - 1),
+            one,
+            &Time::zero(),
+            Some(last_round_reg),
+        );
+        let is_dummy = builder.load(&is_dummy_slice.get_at(process_id), &Time::zero());
+
+        let public = SHAPublicData {
+            initial_hash,
+            padded_chunks: padded_chunks.to_vec(),
+            digest_indices,
+        };
+
+        let trace = SHATraceData {
+            is_preprocessing,
+            process_id,
+            cycle_end_bit,
+            index,
+            is_dummy,
+        };
+
+        let memory = SHAMemory {
+            round_constants,
+            w,
+            shift_read_mult,
+            end_bit,
+            digest_bit,
+            dummy_index,
+        };
+        SHAData {
+            public,
+            trace,
+            memory,
+            degree: 1 << degree_log,
+        }
+    }
+
+    fn preprocessing(
+        builder: &mut B,
+        data: &SHAData<Self::Variable, CYCLE_LENGTH>,
+    ) -> Self::Variable {
+        let w = &data.memory.w;
+        let dummy_index = data.memory.dummy_index;
+        let is_preprocessing = data.trace.is_preprocessing;
+        let shift_read_mult = &data.memory.shift_read_mult;
+        let is_dummy = data.trace.is_dummy;
+        let clk = Self::clk(builder);
+
+        let time = Time::zero();
+
+        let shifted_index = |i: u32, builder: &mut B| {
+            builder.expression(
+                is_dummy.expr() * dummy_index.expr()
+                    + is_dummy.not_expr()
+                        * (is_preprocessing.expr()
+                            * (clk.expr() - B::Field::from_canonical_u32(i))
+                            + is_preprocessing.not_expr() * dummy_index.expr()),
+            )
+        };
+
+        let i_m_15 = shifted_index(15, builder);
+        let w_i_minus_15 = builder.load(&w.get_at(i_m_15), &time);
+        let i_m_2 = shifted_index(2, builder);
+        let w_i_minus_2 = builder.load(&w.get_at(i_m_2), &time);
+
+        let i_m_16 = shifted_index(16, builder);
+        let w_i_mimus_16 = builder.load(&w.get_at(i_m_16), &time);
+        let i_m_7 = shifted_index(7, builder);
+        let w_i_mimus_7 = builder.load(&w.get_at(i_m_7), &time);
+
+        let w_i_pre_process = Self::preprocessing_step(
+            builder,
+            w_i_minus_15,
+            w_i_minus_2,
+            w_i_mimus_16,
+            w_i_mimus_7,
+        );
+
+        let mut i_idx = builder.select(is_preprocessing, &dummy_index, &clk);
```

Comment:
I think a comment here, explaining what's going on is helpful with the loads and stores.

Reply:
I notice you precompute the multiplicites somewhere else as well, it can be useful to have some commentary there.